// bp-space-invaders/esm2022/lib/enum/observable-type.enum.mjs
export var ObservableTypeEnum;
(function (ObservableTypeEnum) {
    ObservableTypeEnum["alienBossDestroyed"] = "alienBossDestroyed";
    ObservableTypeEnum["alienCreepLeftBoundary"] = "alienCreepLeftBoundary";
    ObservableTypeEnum["alienCreepRightBoundary"] = "alienCreepRightBoundary";
    ObservableTypeEnum["alienDestroyed"] = "alienDestroyed";
    ObservableTypeEnum["alienDrop"] = "alienDrop";
    ObservableTypeEnum["bombDropped"] = "bombDropped";
    ObservableTypeEnum["displayScore"] = "displayScore";
    ObservableTypeEnum["gameOver"] = "gameOver";
    ObservableTypeEnum["gamePaused"] = "gamePaused";
    ObservableTypeEnum["gameStarted"] = "gameStarted";
    ObservableTypeEnum["keyDownEvent"] = "keyDownEvent";
    ObservableTypeEnum["keyUpEvent"] = "keyUpEvent";
    ObservableTypeEnum["missileShot"] = "missileShot";
    ObservableTypeEnum["moonDefenseDestroyed"] = "moonDefenseDestroyed";
    ObservableTypeEnum["spaceshipDestroyed"] = "spaceshipDestroyed";
})(ObservableTypeEnum || (ObservableTypeEnum = {}));
