// bp-space-invaders/esm2022/lib/service/game-cookie.service.mjs
import { Injectable } from '@angular/core';
import { CookieService } from 'ngx-cookie-service';
import * as i0 from "@angular/core";
import * as i1 from "ngx-cookie-service";
export class GameCookieService {
    constructor(cookieService) {
        this.cookieService = cookieService;
    }
    getCookie(cookieName) {
        return this.cookieService.get(cookieName);
    }
    getGameData() {
        const cookie = this.getCookie('bp-space-invaders');
        return cookie ? JSON.parse(cookie) : { highScore: 0 };
    }
    setGameData(data) {
        this.cookieService.set('bp-space-invaders', JSON.stringify(data));
    }
    getDebugData() {
        const cookie = this.getCookie('bp-space-invaders-debug');
        return cookie ? JSON.parse(cookie)['assetName'] : '';
    }
    getSecondaryDebugData() {
        const cookie = this.getCookie('bp-space-invaders-debug');
        return cookie ? JSON.parse(cookie)['secondaryAssetName'] : '';
    }
    setDebugData(debugData) {
        const cookie = this.getCookie('bp-space-invaders-debug');
        const existingCookie = (cookie ? JSON.parse(cookie) : {});
        existingCookie.assetName = debugData.assetName ? debugData.assetName : existingCookie.assetName;
        existingCookie.secondaryAssetName = debugData.secondaryAssetName
            ? debugData.secondaryAssetName
            : existingCookie.secondaryAssetName;
        this.cookieService.set('bp-space-invaders-debug', JSON.stringify(existingCookie));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: GameCookieService, deps: [{ token: i1.CookieService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: GameCookieService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: GameCookieService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: i1.CookieService }] });
