// bp-space-invaders/esm2022/lib/service/invasion.service.mjs
import { Alien1Class } from '../aliens/alien-1/alien-1.class';
import { Alien2Class } from '../aliens/alien-2/alien-2.class';
import { Alien3Class } from '../aliens/alien-3/alien-3.class';
import { Alien4Class } from '../aliens/alien-4/alien-4.class';
import { Alien5Class } from '../aliens/alien-5/alien-5.class';
import { Alien6Class } from '../aliens/alien-6/alien-6.class';
import { AlienBoss2Class } from '../aliens/alien-boss-2/alien-boss-2.class';
import { AssetClass } from '../asset/asset.class';
import { AssetAlignVerticalEnum } from '../asset/enum/asset-align-vertical.enum';
import { AssetAlignEnum } from '../asset/enum/asset-align.enum';
import { ObservableTypeEnum } from '../enum/observable-type.enum';
import { ALIEN_COLUMNS, ALIEN_ROWS } from '../gameboard/gameboard-constants';
export class InvasionService extends AssetClass {
    constructor(gameService, ctx, boundary) {
        super(gameService, ctx, boundary, 0);
        this.aliens = [];
        this.isAlienCreepLeft = true;
        this.xAlienDelta = -10;
        this.isBossing = false;
        this.displayElapseTime = 0;
        this.isGameStarted = false;
        this.setup();
        this.gameService.getMasterObservable().subscribe((resultModel) => {
            if (resultModel.type === ObservableTypeEnum.gameStarted) {
                this.isGameStarted = resultModel.booleanData;
                if (resultModel.booleanData === false) {
                    this.alienBoss.isDestroyed = true;
                    this.isBossing = false;
                }
            }
            else if (resultModel.type === ObservableTypeEnum.gamePaused && !resultModel.booleanData) {
                this.setAlienBossDisplayElapseTime();
            }
            else if (resultModel.type === ObservableTypeEnum.alienBossDestroyed) {
                this.toggleScore(true);
                this.isBossing = false;
                this.alienBoss.isDestroyed = true;
                this.setAlienBossDisplayElapseTime();
            }
            else if (resultModel.type === ObservableTypeEnum.alienDestroyed) {
                this.aliens[resultModel.numberData].isDestroyed = true;
                this.aliens[resultModel.numberData].canFireWeapon = false;
                this.findNextBombingAlien(resultModel.numberData);
            }
        });
    }
    spawn() {
        this.assetAnimation = [[0]];
        this.engineAnimation = [];
        super.spawn();
    }
    setup() {
        let alienRowCounter = 0;
        Array(ALIEN_ROWS)
            .fill(0)
            .map(() => {
            Array(ALIEN_COLUMNS)
                .fill(0)
                .map(() => {
                if (alienRowCounter === 1) {
                    this.aliens.push(new Alien2Class(this.gameService, this.ctx, this.boundarySetup, this.aliens.length));
                }
                else if (alienRowCounter === 2) {
                    this.aliens.push(new Alien3Class(this.gameService, this.ctx, this.boundarySetup, this.aliens.length));
                }
                else if (alienRowCounter === 3) {
                    this.aliens.push(new Alien4Class(this.gameService, this.ctx, this.boundarySetup, this.aliens.length));
                }
                else if (alienRowCounter === 4) {
                    this.aliens.push(new Alien5Class(this.gameService, this.ctx, this.boundarySetup, this.aliens.length));
                }
                else if (alienRowCounter === 5) {
                    this.aliens.push(new Alien6Class(this.gameService, this.ctx, this.boundarySetup, this.aliens.length));
                }
                else {
                    this.aliens.push(new Alien1Class(this.gameService, this.ctx, this.boundarySetup, this.aliens.length));
                }
            });
            alienRowCounter++;
        });
        this.alienBoss = new AlienBoss2Class(this.gameService, this.ctx, {
            x: this.boundarySetup.x - 15,
            y: 0,
            height: 20,
            width: this.boundarySetup.width + 20
        }, 0);
        this.aliens.push(this.alienBoss);
    }
    toggleScore(displayScore) {
        this.gameService.emitMasterObservableEvent({
            type: ObservableTypeEnum.displayScore,
            booleanData: displayScore
        });
    }
    findNextBombingAlien(alienPosition) {
        const alienColumnIndex = alienPosition % ALIEN_COLUMNS;
        const startingPosition = this.aliens.length - 1 - (ALIEN_COLUMNS - alienColumnIndex);
        for (let index = startingPosition; index > 0; index -= ALIEN_COLUMNS) {
            if (!this.aliens[index].isDestroyed) {
                this.aliens[index].canFireWeapon = true;
                return;
            }
        }
    }
    aliensDestroyed() {
        return this.getActiveAliens().length === 0;
    }
    setupInvasion(yStart) {
        const xStart = 60;
        const xSpacer = 24;
        const ySpacer = 20;
        let xValue = 0;
        let yValue = 0;
        this.aliens.map((alien) => {
            alien.isDestroyed = false;
            alien.move({
                x: xValue++ * xSpacer + xStart,
                y: yValue * ySpacer + yStart
            });
            alien.canFireWeapon = false;
            if (yValue === ALIEN_ROWS - 1) {
                alien.canFireWeapon = true;
            }
            if (xValue % ALIEN_COLUMNS === 0) {
                yValue++;
                xValue = 0;
            }
        });
        const asset = this.alienBoss.positionAsset(AssetAlignEnum.right, AssetAlignVerticalEnum.top);
        this.alienBoss.move(asset);
    }
    splat() {
        let isGameOver = false;
        this.aliens
            .filter((alien) => !alien.isDestroyed)
            .map((alien) => {
            if (alien.y + alien.getAssetHeight() >= this.boundary.floor) {
                isGameOver = true;
            }
        });
        return isGameOver;
    }
    getAllActiveAliens() {
        return this.aliens.filter((alien) => alien.isDestroyed === false);
    }
    getActiveAliens() {
        return this.aliens.filter((alien) => alien.isDestroyed === false && alien.isAlien());
    }
    increaseInvasionSpeed(creepSpeed) {
        const totalAliens = this.getActiveAliens().length;
        if (totalAliens < 20 && totalAliens > 15 && creepSpeed < 2) {
            return true;
        }
        else if (totalAliens < 15 && totalAliens > 10 && creepSpeed < 3) {
            return true;
        }
        else if (totalAliens < 10 && totalAliens > 5 && creepSpeed < 4) {
            return true;
        }
        else if (totalAliens < 4 && totalAliens > 1 && creepSpeed < 5) {
            return true;
        }
        else if (totalAliens === 1 && creepSpeed < 6) {
            return true;
        }
        return false;
    }
    dropAliens() {
        this.isAlienCreepLeft = !this.isAlienCreepLeft;
        let lastAlien;
        this.getActiveAliens().map((alien) => {
            alien.move({
                x: alien.x,
                y: (alien.y += 10)
            });
            lastAlien = alien;
        });
        this.gameService.emitMasterObservableEvent({
            type: ObservableTypeEnum.alienDrop,
            numberData: lastAlien.y + lastAlien.getAssetHeight()
        });
    }
    moveAliens() {
        let creepSucceeded = true;
        this.getActiveAliens().map((alien) => {
            creepSucceeded =
                creepSucceeded &&
                    alien.creep({
                        x: alien.x + this.xAlienDelta,
                        y: alien.y
                    });
        });
        if (creepSucceeded) {
            this.getActiveAliens().map((alien) => {
                alien.move({
                    x: alien.x + this.xAlienDelta,
                    y: alien.y
                });
            });
        }
        else {
            this.dropAliens();
        }
        this.xAlienDelta = this.isAlienCreepLeft ? -10 : 10;
    }
    animate(now, creepSpeedLevel) {
        this.assetTimer.elapsed = now - this.assetTimer.start;
        if (this.assetTimer.elapsed > creepSpeedLevel) {
            this.assetTimer.start = now;
            this.moveAliens();
        }
        if (this.isBossing) {
            this.alienBoss.animate(now);
        }
        this.getActiveAliens().map((alien) => {
            alien.animate(now);
        });
    }
    resetAlienCreep() {
        this.toggleScore(true);
        this.isBossing = false;
        this.xAlienDelta = -10;
        this.isAlienCreepLeft = true;
        this.setAlienBossDisplayElapseTime();
    }
    setAlienBossDisplayElapseTime() {
        clearTimeout(this.timerId);
        if (!this.isGameStarted || this.isBossing) {
            return;
        }
        this.displayElapseTime = Math.floor(Math.random() * 20000 + 1000);
        this.timerId = window.setTimeout(() => {
            this.isBossing = true;
            this.toggleScore(false);
            this.alienBoss.positionAlienBoss();
        }, this.displayElapseTime);
    }
}
