// bp-space-invaders/esm2022/lib/service/moon.service.mjs
export class MoonService {
    constructor(ctx, boundary) {
        this.ctx = ctx;
        this.boundary = boundary;
        this.leftBoundary = this.boundary.x + 40;
        this.rightBoundary = this.boundary.width - 80 + this.leftBoundary;
        this.boundaryStakeWidth = 5;
        this.boundaryStakeHeight = this.boundary.height / 2;
    }
    drawTheMoon() {
        this.ctx.fillStyle = 'rgba(34, 65, 4)';
        this.ctx.fillRect(this.boundary.x, this.boundary.y, this.boundary.width, this.boundary.height);
        this.ctx.fillStyle = 'rgba(8, 136, 23)';
        this.ctx.fillRect(this.leftBoundary - this.boundaryStakeWidth, this.boundary.y, this.boundaryStakeWidth, this.boundaryStakeHeight);
        this.ctx.fillStyle = 'rgba(200, 95, 36)';
        this.ctx.fillRect(this.rightBoundary, this.boundary.y, this.boundaryStakeWidth, this.boundaryStakeHeight);
    }
    animate() {
        this.drawTheMoon();
    }
    getMoonTop() {
        return this.boundary.y;
    }
    getMoonLeftBoundary() {
        return this.leftBoundary - 1;
    }
    getMoonRightBoundary() {
        return this.rightBoundary + 1;
    }
}
