// bp-space-invaders/esm2022/lib/service/moon-defense.service.mjs
import { MoonDefenseClass } from '../moon-defense/moon-defense.class';
import { AssetAlignEnum } from '../asset/enum/asset-align.enum';
import { AssetAlignVerticalEnum } from '../asset/enum/asset-align-vertical.enum';
import { ObservableTypeEnum } from '../enum/observable-type.enum';
export class MoonDefenseService {
    constructor(gameService, ctx, boundary) {
        this.gameService = gameService;
        this.ctx = ctx;
        this.boundary = boundary;
        this.moonDefenses = [];
        this.setUp();
        this.gameService.getMasterObservable().subscribe((resultModel) => {
            if (resultModel.type === ObservableTypeEnum.moonDefenseDestroyed) {
                this.moonDefenses[resultModel.numberData].isDestroyed = true;
            }
            else if (resultModel.type === ObservableTypeEnum.alienDrop) {
                this.aliensDestroyedMoonbase(resultModel.numberData);
            }
        });
    }
    aliensDestroyedMoonbase(alienYPosition) {
        const moonbase = this.getActiveMoonBases().pop();
        if (moonbase && alienYPosition > moonbase.y) {
            this.getActiveMoonBases().map((moonDefense) => {
                moonDefense.isDestroyed = true;
            });
        }
    }
    setUp() {
        Array(3)
            .fill(0)
            .map(() => {
            this.moonDefenses.push(new MoonDefenseClass(this.gameService, this.ctx, this.boundary, this.moonDefenses.length));
        });
    }
    setupMoonBase(isReset, spaceshipYPosition) {
        this.moonDefenses.map((moonDefense, index) => {
            moonDefense.isDestroyed = false;
            if (isReset) {
                moonDefense.restore();
            }
            let moonAsset = moonDefense.positionAsset(AssetAlignEnum.leftFourth, AssetAlignVerticalEnum.middle);
            if (index === 1) {
                moonAsset = moonDefense.positionAsset(AssetAlignEnum.center, AssetAlignVerticalEnum.middle);
            }
            else if (index === 2) {
                moonAsset = moonDefense.positionAsset(AssetAlignEnum.rightFourth, AssetAlignVerticalEnum.middle);
            }
            moonDefense.move({
                x: moonAsset.x,
                y: spaceshipYPosition - moonDefense.getAssetHeight()
            });
        });
    }
    getActiveMoonBases() {
        return this.moonDefenses.filter((moonDefense) => !moonDefense.isDestroyed);
    }
    animate() {
        this.getActiveMoonBases().map((moonDefense) => {
            moonDefense.animate();
        });
    }
}
