// bp-space-invaders/esm2022/lib/service/text.service.mjs
import { NumberClass } from '../number/number.class';
import { ObservableTypeEnum } from '../enum/observable-type.enum';
export class TextService {
    constructor(gameService, ctx, boundary) {
        this.gameService = gameService;
        this.ctx = ctx;
        this.scoreClass = new NumberClass(this.gameService, this.ctx, boundary, 1);
        this.highScoreClass = new NumberClass(this.gameService, this.ctx, boundary, 2);
        this.isDisplayed = true;
        this.scoreClass.move({
            x: 160,
            y: 0
        });
        this.highScoreClass.move({
            x: 20,
            y: 0
        });
        this.gameService.getMasterObservable().subscribe((resultModel) => {
            if (resultModel.type === ObservableTypeEnum.displayScore) {
                this.isDisplayed = resultModel.booleanData;
            }
            else if (resultModel.type === ObservableTypeEnum.gameStarted) {
                this.isDisplayed = true;
            }
        });
    }
    animate() {
        if (!this.isDisplayed) {
            return;
        }
        this.scoreClass.animate();
        this.highScoreClass.animate();
    }
    setScore(points) {
        this.scoreClass.setNumber(points);
    }
    setHighScore(points) {
        this.highScoreClass.setNumber(points);
    }
    displayText(text, fontColor, leftOffset) {
        const gameboard = this.gameService.getGameboard();
        this.ctx.fillStyle = 'black';
        const gameOverBoxWidth = 100;
        const gameOverBoxHeight = 50;
        const gameOverXCenter = (gameboard.width - gameOverBoxWidth) / 2;
        const gameOverYCenter = (gameboard.height - gameOverBoxHeight) / 2;
        this.ctx.fillRect(gameOverXCenter, gameOverYCenter, gameOverBoxWidth, gameOverBoxHeight);
        this.ctx.font = '10px "Press Start 2P", cursive';
        this.ctx.fillStyle = fontColor;
        this.ctx.fillText(text, gameOverXCenter + leftOffset, gameOverYCenter + 30);
    }
}
