// bp-space-invaders/esm2022/lib/number/number.class.mjs
import { ASSET_ANIMATION, COLORS } from './number-constants';
import { AssetClass } from '../asset/asset.class';
export class NumberClass extends AssetClass {
    constructor(gameService, ctx, boundarySetup, color) {
        super(gameService, ctx, boundarySetup, 0);
        this.color = color;
        this.numbers = [];
    }
    spawn() {
        this.assetAnimation = ASSET_ANIMATION[0];
        this.engineAnimation = [];
        this.destructionAnimations = [];
        super.spawn();
    }
    draw() {
        this.ctx.save();
        this.ctx.translate(this.x, this.y);
        this.ctx.scale(2, 2);
        this.drawNumber();
        this.ctx.restore();
    }
    drawNumber() {
        let offset = 0;
        this.ctx.fillStyle = COLORS[this.color];
        this.numbers.map((scoreNumber) => {
            this.assetAnimation = ASSET_ANIMATION[scoreNumber];
            this.assetAnimation.forEach((row, y) => {
                row.forEach((value, x) => {
                    if (value > 0) {
                        this.ctx.fillRect(x + offset, y, 1, 1);
                    }
                });
            });
            offset += 10;
        });
    }
    padPoints(points) {
        let stringPoints = points.toString();
        if (points < 10) {
            stringPoints = `000${points}`;
        }
        else if (points < 100) {
            stringPoints = `00${points}`;
        }
        else if (points < 1000) {
            stringPoints = `0${points}`;
        }
        return stringPoints;
    }
    setNumber(inputNumber, isPadNumber = true) {
        const numbers = isPadNumber ? this.padPoints(inputNumber) : [inputNumber];
        this.numbers = [];
        for (const index of numbers) {
            this.numbers.push(Number(index));
        }
    }
    animate() {
        this.draw();
    }
}
