// bp-space-invaders/esm2022/lib/aliens/alien/alien-constants.mjs
export const COLS = 31;
export const ROWS = 31;
export const BLOCK_SIZE = 25;
export const COLORS = [
    'none',
    'rgba(200, 95, 36)',
    'rgba(255, 0, 0)',
    'rgba(122, 122, 122)',
    'rgba(163, 46, 132)'
];
export class KEY {
    static { this.DOWN = 'ArrowDown'; }
    static { this.LEFT = 'ArrowLeft'; }
    static { this.RIGHT = 'ArrowRight'; }
}
const DESTRUCTION_ANIMATION_1 = [
    [2, 0, 0, 0, 0, 0, 0, 0, 0, 2],
    [0, 2, 0, 0, 0, 0, 0, 0, 2, 0],
    [0, 0, 2, 3, 2, 0, 2, 0, 0, 0],
    [0, 0, 2, 0, 2, 2, 3, 2, 0, 0],
    [0, 0, 0, 2, 3, 2, 2, 0, 0, 0],
    [0, 0, 2, 3, 2, 2, 0, 2, 0, 0],
    [0, 0, 2, 0, 2, 3, 2, 2, 0, 0],
    [0, 2, 0, 0, 0, 0, 0, 0, 2, 0],
    [2, 0, 0, 0, 0, 0, 0, 0, 0, 2]
];
const DESTRUCTION_ANIMATION_2 = [
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 2, 0, 0, 0, 0, 2, 0, 0],
    [0, 0, 0, 3, 0, 2, 3, 0, 0, 0],
    [0, 0, 0, 0, 3, 0, 0, 0, 0, 0],
    [0, 0, 0, 3, 0, 2, 2, 0, 0, 0],
    [0, 0, 2, 0, 0, 0, 0, 2, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
];
const DESTRUCTION_ANIMATION_3 = [
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 3, 0, 2, 0, 0, 0, 0],
    [0, 0, 0, 0, 3, 0, 0, 0, 0, 0],
    [0, 0, 0, 2, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
];
export const DESTRUCTION_ANIMATION = [DESTRUCTION_ANIMATION_1, DESTRUCTION_ANIMATION_2, DESTRUCTION_ANIMATION_3];
