// bp-space-invaders/esm2022/lib/missile/missile/test-component/missile.component.mjs
import { Component, ElementRef, ViewChild } from '@angular/core';
import { BLOCK_SIZE, COLS, ROWS } from '../missile-constants';
import { GameService } from '../../../service/game.service';
import { AssetAlignEnum } from '../../../asset/enum/asset-align.enum';
import { AssetAlignVerticalEnum } from '../../../asset/enum/asset-align-vertical.enum';
import { Missile1Class } from '../../missile-1/missile-1.class';
import * as i0 from "@angular/core";
import * as i1 from "../../../service/game.service";
export class MissileComponent {
    constructor(gameService) {
        this.gameService = gameService;
    }
    ngOnInit() {
        this.ctx = this.gameService.initBoard(this.canvas, COLS, ROWS, BLOCK_SIZE);
        const boundary = {
            x: 0,
            y: 0,
            height: ROWS,
            width: COLS
        };
        this.missile = new Missile1Class(this.gameService, this.ctx, boundary, 0);
        this.missile.move(this.missile.positionAsset(AssetAlignEnum.center, AssetAlignVerticalEnum.bottom));
        this.isAnimateOnce = true;
        this.animate();
    }
    animateAsset() {
        this.isAnimateOnce = !this.isAnimateOnce;
        this.missile.isDestroyed = false;
        this.missile.move(this.missile.positionAsset(AssetAlignEnum.center, AssetAlignVerticalEnum.middle));
        cancelAnimationFrame(this.requestId);
        this.animate();
    }
    draw() {
        this.ctx.clearRect(0, 0, this.ctx.canvas.width, this.ctx.canvas.height);
        this.gameService.addOutlines(this.ctx, 1, 0.025);
    }
    animate(now = 0) {
        this.draw();
        this.missile.animate(now);
        if (!this.isAnimateOnce) {
            this.requestId = requestAnimationFrame(this.animate.bind(this));
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MissileComponent, deps: [{ token: i1.GameService }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "17.0.7", type: MissileComponent, selector: "bp-missile", viewQueries: [{ propertyName: "canvas", first: true, predicate: ["board"], descendants: true, static: true }], ngImport: i0, template: "<div class=\"game-container\">\n  Missile\n  <div class=\"game-board-container\">\n    <canvas #board class=\"game-board\"></canvas>\n  </div>\n  <div class=\"button-container\">\n    <button (click)=\"animateAsset()\" class=\"animation-button button\">Animate</button>\n  </div>\n</div>\n", styles: [".game-container .header{display:grid;grid-template-columns:200px auto 200px;height:25px}.game-container .header .game-title{text-align:center;border:1px solid black}.game-container .game-board-container{display:flex;flex-direction:row;justify-content:center}.game-container .game-board-container .game-board{border:solid 2px grey;background-color:#000}.game-container .button-container{position:relative;margin-top:25px;display:flex;flex-direction:row;justify-content:center}.game-container .button-container .animation-button{background-color:#4caf50}.game-container .button-container .button{font-size:16px;padding:15px 30px;cursor:pointer;width:165px;border:1px solid black}\n"] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: MissileComponent, decorators: [{
            type: Component,
            args: [{ selector: 'bp-missile', template: "<div class=\"game-container\">\n  Missile\n  <div class=\"game-board-container\">\n    <canvas #board class=\"game-board\"></canvas>\n  </div>\n  <div class=\"button-container\">\n    <button (click)=\"animateAsset()\" class=\"animation-button button\">Animate</button>\n  </div>\n</div>\n", styles: [".game-container .header{display:grid;grid-template-columns:200px auto 200px;height:25px}.game-container .header .game-title{text-align:center;border:1px solid black}.game-container .game-board-container{display:flex;flex-direction:row;justify-content:center}.game-container .game-board-container .game-board{border:solid 2px grey;background-color:#000}.game-container .button-container{position:relative;margin-top:25px;display:flex;flex-direction:row;justify-content:center}.game-container .button-container .animation-button{background-color:#4caf50}.game-container .button-container .button{font-size:16px;padding:15px 30px;cursor:pointer;width:165px;border:1px solid black}\n"] }]
        }], ctorParameters: () => [{ type: i1.GameService }], propDecorators: { canvas: [{
                type: ViewChild,
                args: ['board', { static: true }]
            }] } });
