// bp-space-invaders/esm2022/lib/missile/missile/missile.class.mjs
import { COLORS, KEY } from './missile-constants';
import { AssetClass } from '../../asset/asset.class';
export class MissileClass extends AssetClass {
    constructor(gameService, ctx, boundarySetup, guid) {
        super(gameService, ctx, boundarySetup, guid);
        this.moves = {
            [KEY.UP]: (asset) => ({ ...asset, y: asset.y - 5 })
        };
    }
    spawn() {
        this.assetAnimation = this.assetAnimations[0];
        this.engineAnimation = this.engineAnimations[0];
        super.spawn();
    }
    draw() {
        this.ctx.save();
        this.ctx.translate(this.x, this.y);
        this.drawMissile();
        this.drawEngine();
        this.ctx.restore();
    }
    drawMissile() {
        this.assetAnimation.forEach((row, y) => {
            row.forEach((value, x) => {
                this.ctx.fillStyle = COLORS[value];
                this.ctx.fillRect(x, y, 1, 1);
            });
        });
    }
    drawEngine() {
        this.engineAnimation.forEach((row, y) => {
            row.forEach((value, x) => {
                this.ctx.fillStyle = COLORS[value];
                this.ctx.fillRect(this.x + this.engineXOffset + x, this.y + this.engineYOffset + y, 1, 1);
            });
        });
    }
    animate(now = 0) {
        this.engineTimer.elapsed = now - this.engineTimer.start;
        if (this.engineTimer.elapsed > 30) {
            this.engineTimer.start = now;
            this.engineAnimation = this.engineAnimations[this.engineAnimationCounter++ % this.engineAnimations.length];
            const asset = this.moves['ArrowUp'](this);
            if (this.gameService.valid(asset, this.boundary)) {
                this.move(asset);
            }
            else {
                this.isDestroyed = true;
            }
        }
        this.draw();
    }
}
