// bp-space-invaders/esm2022/lib/space-invaders.module.mjs
import { NgModule } from '@angular/core';
import { SpaceInvadersComponent } from './space-invaders/space-invaders.component';
import { GameboardComponent } from './gameboard/gameboard.component';
import { GameService } from './service/game.service';
import { CommonModule } from '@angular/common';
import { AlienComponent } from './aliens/alien/test-component/alien.component';
import { SpaceshipComponent } from './spaceships/spaceship/test-component/spaceship.component';
import { MissileComponent } from './missile/missile/test-component/missile.component';
import { BombComponent } from './bomb/test-component/bomb.component';
import { MoonDefenseComponent } from './moon-defense/test-component/moon-defense.component';
import { AlienBossComponent } from './aliens/alien-boss/test-component/alien-boss.component';
import { NumberComponent } from './number/test-component/number.component';
import { GameCookieService } from './service/game-cookie.service';
import { CookieService } from 'ngx-cookie-service';
import * as i0 from "@angular/core";
export class SpaceInvadersModule {
    static forRoot(gameConfiguration) {
        return {
            ngModule: SpaceInvadersModule,
            providers: [
                {
                    provide: 'GameConfigurationData',
                    useValue: gameConfiguration
                },
                CookieService,
                GameCookieService
            ]
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: SpaceInvadersModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.0.7", ngImport: i0, type: SpaceInvadersModule, declarations: [AlienBossComponent,
            AlienComponent,
            BombComponent,
            GameboardComponent,
            MissileComponent,
            MoonDefenseComponent,
            NumberComponent,
            SpaceInvadersComponent,
            SpaceshipComponent], imports: [CommonModule], exports: [SpaceInvadersComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: SpaceInvadersModule, providers: [CookieService, GameCookieService, GameService], imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.7", ngImport: i0, type: SpaceInvadersModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [
                        AlienBossComponent,
                        AlienComponent,
                        BombComponent,
                        GameboardComponent,
                        MissileComponent,
                        MoonDefenseComponent,
                        NumberComponent,
                        SpaceInvadersComponent,
                        SpaceshipComponent
                    ],
                    exports: [SpaceInvadersComponent],
                    providers: [CookieService, GameCookieService, GameService]
                }]
        }] });
