// bp-space-invaders/esm2022/lib/aliens/alien-boss/alien-boss.class.mjs
import { AlienClass } from '../alien/alien.class';
import { KEY } from '../alien/alien-constants';
import { ObservableTypeEnum } from '../../enum/observable-type.enum';
import { AssetAlignEnum } from '../../asset/enum/asset-align.enum';
import { AssetAlignVerticalEnum } from '../../asset/enum/asset-align-vertical.enum';
export class AlienBossClass extends AlienClass {
    constructor(gameService, ctx, boundarySetup, guid, points) {
        super(gameService, ctx, boundarySetup, guid, points);
        this.moves = {
            [KEY.LEFT]: (asset) => ({ ...asset, x: asset.x - 5 }),
            [KEY.RIGHT]: (asset) => ({ ...asset, x: asset.x + 5 })
        };
        this.directionLeft = true;
        this.isAlienAsset = false;
    }
    determineBossDirection() {
        this.directionLeft = Math.floor(Math.random() * 10) + 1 > 5;
    }
    positionAlienBoss() {
        this.determineBossDirection();
        const direction = this.directionLeft ? AssetAlignEnum.right : AssetAlignEnum.left;
        const asset = this.positionAsset(direction, AssetAlignVerticalEnum.top);
        asset.y = this.y;
        this.move(asset);
        this.isDestroyed = false;
    }
    animate(now = 0) {
        this.assetTimer.elapsed = now - this.assetTimer.start;
        if (this.isDestroyed) {
            this.destructionAnimate();
            return;
        }
        if (this.isGameOver) {
            this.isDestroyed = true;
            return;
        }
        this.destructionAnimationCounter = 0;
        if (this.assetTimer.elapsed > 200) {
            this.assetTimer.start = now;
            this.assetAnimation = this.assetAnimations[this.animationCounter++ % this.assetAnimations.length];
            const asset = this.directionLeft ? this.moves['ArrowLeft'](this) : this.moves['ArrowRight'](this);
            if (this.gameService.valid(asset, this.boundary)) {
                this.move(asset);
            }
            else {
                this.isDestroyed = true;
                this.gameService.emitMasterObservableEvent({
                    type: ObservableTypeEnum.alienBossDestroyed,
                    booleanData: true,
                    pointData: 0
                });
            }
        }
        this.draw();
    }
}
